src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -0,0 +1,318 @@
+import { getDisplayPath } from '../../platform/common/platform/fs-paths';
+import { IDisposableRegistry, Resource } from '../../platform/common/types';
+import { getNotebookMetadata, getResourceType } from '../../platform/common/utils';
+import { noop } from '../../platform/common/utils/misc';
+import { IInterpreterService } from '../../platform/interpreter/contracts';
+import { traceError, traceInfo, traceInfoIfCI, traceVerbose } from '../../platform/logging';
+import { PythonEnvironment } from '../../platform/pythonEnvironments/info';
+import { sendTelemetryEvent } from '../../telemetry';
+import { getTelemetrySafeLanguage } from '../../telemetry/helpers';
+import {
+    IControllerDefaultService,
+    IControllerLoader,
+    IControllerPreferredService,
+    IControllerRegistration,
+    IVSCodeNotebookController,
+    PreferredKernelExactMatchReason
+} from './types';
+
+/**
+ * This class keeps track of selected controllers
+ */
+@injectable()
+export class ControllerPreferredService implements IControllerPreferredService, IExtensionSingleActivationService {
+    private preferredControllers = new Map<NotebookDocument, IVSCodeNotebookController>();
+    private get isLocalLaunch(): boolean {
+        return this.serverConnectionType.isLocalLaunch;
+    }
+    constructor(
+        @inject(IControllerRegistration) private readonly registration: IControllerRegistration,
+        @inject(IControllerLoader) private readonly loader: IControllerLoader,
+        @inject(IControllerDefaultService) private readonly defaultService: IControllerDefaultService,
+        @inject(IInterpreterService) private readonly interpreters: IInterpreterService,
+        @inject(IVSCodeNotebook) private readonly notebook: IVSCodeNotebook,
+        @inject(IDisposableRegistry) readonly disposables: IDisposableRegistry,
+        @inject(IKernelFinder) private readonly kernelFinder: IKernelFinder,
+        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker,
+        @inject(ServerConnectionType) private readonly serverConnectionType: ServerConnectionType
+    ) {}
+    public async activate() {
+        // Sign up for document either opening or closing
+        this.notebook.onDidOpenNotebookDocument(this.onDidOpenNotebookDocument, this, this.disposables);
+        // If the extension activates after installing Jupyter extension, then ensure we load controllers right now.
+        this.notebook.notebookDocuments.forEach((notebook) => this.onDidOpenNotebookDocument(notebook));
+    }
+    public async computePreferred(
+        document: NotebookDocument,
+        serverId?: string | undefined
+    ): Promise<{
+        preferredConnection?: KernelConnectionMetadata | undefined;
+        controller?: IVSCodeNotebookController | undefined;
+    }> {
+        traceInfoIfCI(`Clear controller mapping for ${getDisplayPath(document.uri)}`);
+        const loadControllersPromise = this.loader.loaded;
```

Comment:
Do we need this variable, why not just remove this variable?

Reply:
It's used below.

Reply:
Else it gives the impression that it could change at a later point hence the need to capture into a variable.

Reply:
Maybe before it was so it was this specific instance of loading? That's what it does. It's only going to wait for loading that was going on when this function started. If a refresh happens before we get to the bottom, that refresh doesn't have to be waited on. But maybe we want to wait on any refreshes too.

Reply:
Make sense, thought I do'nt think we'd need to wait on the refreshes, because that would happen before the call to this method.
I.e. i'd expect the calling code to orchestrate that behavior, e.g. refresh, & then call this method - not call this method & then call refresh.

I'm not fussed either way, marking this comment of mine as resolved.